File: backend/app/schemas/quality.py
from pydantic import BaseModel, Field
from typing import Any, Optional, List
from datetime import datetime

from app.models.quality import IssueSeverity


class QualityIssueOut(BaseModel):
    id: int
    report_id: int
    severity: str
    section_ref: Optional[str] = None
    message: str
    suggestion: Optional[str] = None

    model_config = {"from_attributes": True}


class BrokenLinkOut(BaseModel):
    id: int
    report_id: int
    url: str
    status_code: Optional[int] = None
    section_ref: Optional[str] = None

    model_config = {"from_attributes": True}


class QualityReportOut(BaseModel):
    id: int
    document_id: int
    overall_score: float
    completeness: float
    acceptance_coverage: float
    consistency: float
    readability: float
    accuracy: float
    generated_at: datetime
    issues: List[QualityIssueOut] = []
    broken_links: List[BrokenLinkOut] = []
    findings: List["QualityFindingOut"] = []

    model_config = {"from_attributes": True}


class QualityRunResponse(BaseModel):
    message: str
    task_id: str


class QualityReportSummary(BaseModel):
    id: int
    document_id: int
    overall_score: float
    generated_at: datetime

    model_config = {"from_attributes": True}


class QualityStatusResponse(BaseModel):
    status: str
    task_id: str
    message: str
    report: Optional[QualityReportSummary] = None
    error: Optional[str] = None


class QualityFindingOut(BaseModel):
    id: int
    document_id: int
    report_id: Optional[int] = None
    category: str
    status: str
    severity: str
    section_id: Optional[int] = None
    section_ref: Optional[str] = None
    message: str
    suggestion: Optional[str] = None
    quote: Optional[str] = None
    offset: Optional[int] = None
    length: Optional[int] = None
    replacements: list[str] = Field(default_factory=list)
    rule_id: Optional[str] = None
    content_fingerprint: str
    provider: Optional[str] = None
    provider_metadata: Optional[dict[str, Any]] = None
    stale_location: bool = False
    first_seen_at: datetime
    last_seen_at: datetime

    model_config = {"from_attributes": True}


class QualityFindingStatusUpdate(BaseModel):
    status: str


class QualityGrammarRunRequest(BaseModel):
    section_id: Optional[int] = None
    language: str = "en-US"


class QualityAIFixRequest(BaseModel):
    finding_id: Optional[int] = None
    category: Optional[str] = None
    section_id: Optional[int] = None
    status: Optional[str] = "open"
    action: str = "fix_findings"

File: frontend/src/api/quality.ts
import apiClient from './client';
import type { QualityReport } from '@/types';

export interface QualityIssue {
  id: number;
  report_id: number;
  severity: 'error' | 'warning' | 'info';
  section_ref?: string;
  message: string;
  suggestion?: string;
}

export type QualityFindingCategory =
  | 'completeness'
  | 'acceptance'
  | 'terminology'
  | 'links'
  | 'readability'
  | 'grammar'
  | 'accuracy';

export type QualityFindingStatus = 'open' | 'proposed' | 'resolved' | 'dismissed';

export interface QualityFinding {
  id: number;
  document_id: number;
  report_id?: number | null;
  category: QualityFindingCategory;
  status: QualityFindingStatus;
  severity: 'error' | 'warning' | 'info';
  section_id?: number | null;
  section_ref?: string | null;
  message: string;
  suggestion?: string | null;
  quote?: string | null;
  offset?: number | null;
  length?: number | null;
  replacements: string[];
  rule_id?: string | null;
  content_fingerprint: string;
  provider?: string | null;
  provider_metadata?: Record<string, unknown> | null;
  stale_location: boolean;
  first_seen_at: string;
  last_seen_at: string;
}

export interface BrokenLink {
  id: number;
  report_id: number;
  url: string;
  status_code?: number;
  section_ref?: string;
}

export interface QualityReportFull extends QualityReport {
  issues: QualityIssue[];
  broken_links: BrokenLink[];
  findings?: QualityFinding[];
}

export interface QualityRunResponse {
  message: string;
  task_id: string;
}

export interface QualityStatus {
  status: 'queued' | 'running' | 'completed' | 'failed' | 'missing_report';
  task_id: string;
  message: string;
  report?: {
    id: number;
    document_id: number;
    overall_score: number;
    generated_at: string;
  } | null;
  error?: string | null;
}

export const qualityApi = {
  /** Dispatch quality analysis job → 202 */
  async runQuality(projectId: number, documentId: number): Promise<QualityRunResponse> {
    const { data } = await apiClient.post(`/projects/${projectId}/documents/${documentId}/quality/run`);
    return data;
  },

  async getStatus(projectId: number, documentId: number, taskId: string): Promise<QualityStatus> {
    const { data } = await apiClient.get(
      `/projects/${projectId}/documents/${documentId}/quality/status`,
      { params: { task_id: taskId } },
    );
    return data;
  },

  /** Fetch latest quality report with all issues + broken links */
  async getQuality(projectId: number, documentId: number): Promise<QualityReportFull> {
    const { data } = await apiClient.get(`/projects/${projectId}/documents/${documentId}/quality`);
    return data;
  },

  /** Fetch issues, optionally filtered by severity */
  async getIssues(
    projectId: number,
    documentId: number,
    severity?: 'error' | 'warning' | 'info',
  ): Promise<QualityIssue[]> {
    const params = severity ? { severity } : {};
    const { data } = await apiClient.get(`/projects/${projectId}/documents/${documentId}/quality/issues`, { params });
    return data;
  },

  async getFindings(
    projectId: number,
    documentId: number,
    params?: { category?: QualityFindingCategory; status?: QualityFindingStatus; section_id?: number },
  ): Promise<QualityFinding[]> {
    const { data } = await apiClient.get(`/projects/${projectId}/documents/${documentId}/quality/findings`, { params });
    return data;
  },

  async updateFindingStatus(
    projectId: number,
    documentId: number,
    findingId: number,
    status: QualityFindingStatus,
  ): Promise<QualityFinding> {
    const { data } = await apiClient.patch(`/projects/${projectId}/documents/${documentId}/quality/findings/${findingId}`, { status });
    return data;
  },

  async runGrammarFindings(
    projectId: number,
    documentId: number,
    payload: { section_id?: number; language?: string } = {},
  ): Promise<QualityFinding[]> {
    const { data } = await apiClient.post(`/projects/${projectId}/documents/${documentId}/quality/grammar/run`, payload);
    return data;
  },

  async createAiFix(
    projectId: number,
    documentId: number,
    payload: {
      finding_id?: number;
      category?: QualityFindingCategory;
      section_id?: number;
      status?: QualityFindingStatus;
      action?: string;
    },
  ) {
    const { data } = await apiClient.post(`/projects/${projectId}/documents/${documentId}/quality/ai-fix`, payload);
    return data;
  },
};
